client-lite/src/trace/do_log.cpp

```diff
@@ -0,0 +1,437 @@
+#include "do_common.h"
+#include "do_log.h"
+
+#include <dirent.h>
+#include <unistd.h> // getpid
+#include <sys/syscall.h> // SYS_gettid
+#include <cstdarg> // va_start, etc.
+#include <condition_variable>
+#include <fstream>
+#include <mutex>
+#include <string>
+#include <thread>
+#include "do_date_time.h"
+#include "string_ops.h"
+
+namespace DOLog
+{
+
+const char* const g_logFileNamePrefix = "do-agent.";
+constexpr uint32_t g_maxLogFileSizeBytes = 256 * 1024;
+constexpr uint32_t g_maxLogFiles = 3;
+constexpr uint32_t g_logBufferSizeBytes = 64 * 1024;
+
+// Heap-based buffer where log messages are stored before flushing to disk
+class LogBuffer
+{
+public:
+    struct Stats
+    {
+        uint32_t numResets;
+        uint32_t numFlushes;
+        uint32_t numWritesSucceeded;
+        uint32_t numWritesFailed;
+    };
+
+private:
+    std::vector<char> _buffer;
+    size_t _cursor;
+    Stats _stats {};
+
+public:
+    LogBuffer(size_t bufSize) :
+        _buffer(bufSize)
+    {
+        Reset();
+    }
+
+    void Reset()
+    {
+        memset(_buffer.data(), 0, _buffer.size());
+        _cursor = 0;
+        ++_stats.numResets;
+    }
+
+    HRESULT Flush(std::ofstream& fileStream) try
+    {
+        if (fileStream.is_open() && BytesBuffered())
+        {
+            fileStream.write(_buffer.data(), _cursor);
+            fileStream.flush();
+            Reset();
+            ++_stats.numFlushes;
+        }
+        return S_OK;
+    } CATCH_RETURN()
+
+    size_t BytesBuffered() const noexcept
+    {
+        return _cursor;
+    }
+
+    size_t BytesFree() const noexcept
+    {
+        return (_buffer.size() - _cursor);
+    }
+
+    size_t BytesFreePercentage() const noexcept
+    {
+        return BytesFree() * static_cast<uint64_t>(100) / _buffer.size();
+    }
+
+    const Stats& PerfStats() const noexcept
+    {
+        return _stats;
+    }
+
+    HRESULT Write(const char* pszLevel, const char* pszFunc, uint32_t nLine, HRESULT hrIn, const char* pszFmt, va_list argList)
+    {
+        static const pid_t pid = getpid();
+        const auto tid = static_cast<int>(syscall(SYS_gettid)); // using syscall because glibc wrapper is unavailable
+        const auto timeStr = SysTimePointToUTCString(wall_clock_t::now());
+
+        int cchWritten = 0;
+        int cchTotalWritten = 0;
+
+        char* pszBuffer = _buffer.data() + _cursor;
+        const size_t cchBuffer = BytesFree();
+
+        // Timestamp ProcessID ThreadID severity
+        HRESULT hr = StringPrintf(pszBuffer, cchBuffer, &cchWritten, "%s %-5d %-5d %-8s ", timeStr.data(), pid, tid, pszLevel);
+        if (SUCCEEDED(hr))
+        {
+            cchTotalWritten += cchWritten;
+
+            // Note: pszFunc may not be available. Example is logging from error_macros.cpp in release builds.
+            if (hrIn != HRESULT(-1))
```

Comment:
>-1 [](http://example.com/codeflow?start=32&length=2)

S_OK is 0 no?

Reply:
-1 here indicates there is no HRESULT to log. See the calls to this method in Write() and WriteResult() below.